File: year2017/day04/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return sum(len(set(passphrase)) == len(passphrase) for passphrase in input)


def part_2():
  return sum(len(set([''.join(sorted(word)) for word in [*passphrase]])) == len(passphrase) for passphrase in input)


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row.split(' ') for row in file.read().split('\n')]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2017/day04/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static List<String[]> PASSPHRASES = new ArrayList<String[]>();

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        int sum = 0;
        for(String[] passphrase : PASSPHRASES) {
            if (new HashSet<>(Arrays.asList(passphrase)).size() == passphrase.length) {
                sum++;
            }
        }
        return sum;
    }
    
    public static int part2() {
        int sum = 0;
        for(String[] passphrase : PASSPHRASES) {
            for(int i = 0; i < passphrase.length; i++) {
                char[] charArray = passphrase[i].toCharArray();
                Arrays.sort(charArray);
                passphrase[i] = new String(charArray);
            }
            if (new HashSet<>(Arrays.asList(passphrase)).size() == passphrase.length) {
                sum++;
            }
        }
        return sum;
    }

    public static void setup(String input) {
        for(String row : input.split("\n")) {
            PASSPHRASES.add(row.split(" "));
        }
    }
}